File: latest/week2/workshop4_m10/W04_M10_demo.R
# ============================================================
# Workshop 4 — Module 10: Lab — Full EDA on Netflix
# Integrated R Script — UNYT Tirana
# ============================================================
library(tidyverse); library(patchwork); dir.create("output",showWarnings=FALSE)

nf <- read_csv("netflix.csv") |>
  mutate(date_added=mdy(str_trim(date_added)), year_added=year(date_added),
    dur_min=ifelse(type=="Movie",parse_number(duration),NA),
    seasons=ifelse(type=="TV Show",parse_number(duration),NA),
    primary_genre=str_trim(str_extract(listed_in,"^[^,]+")),
    primary_country=str_trim(str_extract(country,"^[^,]+")))
cat("Rows:", nrow(nf), "| Movies:", sum(nf$type=="Movie"), "| TV:", sum(nf$type=="TV Show"), "\n")

# Missingness
cat("\n── Missing % ──\n"); print(round(colMeans(is.na(nf))*100,1))

# Genre explosion
nf_genres <- nf |> separate_rows(listed_in, sep=", ") |> mutate(genre=str_trim(listed_in))

# Dashboard
p_miss <- tibble(col=names(nf),pct=colMeans(is.na(nf))*100) |> filter(pct>0) |>
  mutate(col=fct_reorder(col,pct)) |>
  ggplot(aes(x=col,y=pct)) + geom_col(fill="#E53935",width=0.5) + coord_flip() +
  geom_hline(yintercept=5,linetype="dashed",color="#888") +
  theme_minimal(base_size=8) + labs(title="(a) Missing %",y="%")

p_line <- nf |> filter(release_year>=2000) |> count(release_year,type) |>
  ggplot(aes(x=release_year,y=n,color=type)) + geom_line(linewidth=1) + geom_point(size=1.5) +
  scale_color_manual(values=c(Movie="#1565C0","TV Show"="#E53935")) +
  theme_minimal(base_size=8) + theme(legend.position="bottom") + labs(title="(b) Releases/Year")

p_genre <- nf_genres |> count(genre,sort=TRUE) |> slice_max(n,n=10) |> mutate(genre=fct_reorder(genre,n)) |>
  ggplot(aes(x=genre,y=n)) + geom_col(fill="#E53935",width=0.5) + coord_flip() +
  theme_minimal(base_size=8) + labs(title="(c) Top Genres",x=NULL)

p_dur <- nf |> filter(type=="Movie",!is.na(dur_min)) |>
  ggplot(aes(x=dur_min)) + geom_histogram(bins=40,fill="#1565C0",color="white") +
  geom_vline(xintercept=mean(nf$dur_min,na.rm=T),color="#E53935",linetype="dashed") +
  theme_minimal(base_size=8) + labs(title="(d) Movie Duration")

p_rating <- nf |> count(rating,sort=TRUE) |> slice_max(n,n=8) |> mutate(rating=fct_reorder(rating,n)) |>
  ggplot(aes(x=rating,y=n)) + geom_col(fill="#7B1FA2",width=0.5) + coord_flip() +
  theme_minimal(base_size=8) + labs(title="(e) Content Rating",x=NULL)

p_country <- nf |> count(primary_country,sort=TRUE) |> drop_na() |> slice_max(n,n=8) |>
  mutate(primary_country=fct_reorder(primary_country,n)) |>
  ggplot(aes(x=primary_country,y=n)) +
  geom_segment(aes(xend=primary_country,y=0,yend=n),color="grey75") +
  geom_point(color="#1565C0",size=2) + coord_flip() +
  theme_minimal(base_size=8) + labs(title="(f) Top Countries",x=NULL)

dashboard <- (p_miss|p_line|p_genre) / (p_dur|p_rating|p_country) +
  plot_annotation(title=paste0("Netflix EDA: ",nrow(nf)," Titles"),
    caption="Source: Kaggle | UNYT", tag_levels="a") & theme_minimal(base_size=8)
ggsave("output/W04_M10_netflix.pdf",dashboard,width=14,height=9)
ggsave("output/W04_M10_netflix.png",dashboard,width=14,height=9,dpi=300)
cat("\n── W04-M10 Lab Complete (R) ──\n")

File: latest/week2/workshop4_m10/W04_M10_demo.py
"""W04-M10: Lab — Full EDA on Netflix — Python — UNYT"""
import numpy as np, pandas as pd, matplotlib.pyplot as plt, os
from matplotlib.gridspec import GridSpec
os.makedirs("output",exist_ok=True); np.random.seed(42)

nf=pd.read_csv("netflix.csv")
nf["date_added"]=pd.to_datetime(nf["date_added"].str.strip(),errors="coerce")
nf["year_added"]=nf["date_added"].dt.year
nf["dur_min"]=nf.loc[nf["type"]=="Movie","duration"].str.extract(r"(\d+)").astype(float)
nf["seasons"]=nf.loc[nf["type"]=="TV Show","duration"].str.extract(r"(\d+)").astype(float)
nf["primary_genre"]=nf["listed_in"].str.split(",").str[0].str.strip()
nf["primary_country"]=nf["country"].str.split(",").str[0].str.strip()
print(f"Rows: {len(nf)} | Movies: {(nf['type']=='Movie').sum()} | TV: {(nf['type']=='TV Show').sum()}")
print(f"\nMissing %:\n{(nf.isna().mean()*100).round(1).sort_values(ascending=False)}")

# Dashboard (6 panels)
fig=plt.figure(figsize=(14,9)); gs=GridSpec(2,3,figure=fig,hspace=0.4,wspace=0.35)

ax=fig.add_subplot(gs[0,0]); miss=(nf.isna().mean()*100).sort_values(); miss[miss>0].plot.barh(ax=ax,color="#E53935")
ax.axvline(x=5,color="#888",lw=0.8,ls="--"); ax.set_title("(a) Missing %",fontsize=9,fontweight="bold")

ax=fig.add_subplot(gs[0,1]); yr=nf.query("release_year>=2000").groupby(["release_year","type"]).size().unstack(fill_value=0)
if "Movie" in yr: ax.plot(yr.index,yr["Movie"],"-o",color="#1565C0",lw=1.5,markersize=3,label="Movie")
if "TV Show" in yr: ax.plot(yr.index,yr["TV Show"],"-s",color="#E53935",lw=1.5,markersize=3,label="TV")
ax.legend(fontsize=5); ax.set_title("(b) Releases/Year",fontsize=9,fontweight="bold")

ax=fig.add_subplot(gs[0,2]); genres=nf["listed_in"].str.split(",").explode().str.strip().value_counts().head(10).sort_values()
cl=["#BBBBBB"]*10; cl[-1]="#E53935"; ax.barh(genres.index,genres.values,color=cl,height=0.55)
ax.set_title("(c) Top Genres",fontsize=9,fontweight="bold")
for sp in ["top","right","left"]: ax.spines[sp].set_visible(False)
ax.set_xticks([]); ax.tick_params(left=False,labelsize=5)

ax=fig.add_subplot(gs[1,0]); dur=nf["dur_min"].dropna()
ax.hist(dur,bins=40,color="#1565C0",edgecolor="white",lw=0.3)
ax.axvline(x=dur.mean(),color="#E53935",lw=1.5,ls="--"); ax.set_title("(d) Movie Duration",fontsize=9,fontweight="bold")

ax=fig.add_subplot(gs[1,1]); rats=nf["rating"].value_counts().head(8).sort_values()
ax.barh(rats.index,rats.values,color="#7B1FA2",height=0.55)
ax.set_title("(e) Content Rating",fontsize=9,fontweight="bold")
for sp in ["top","right","left"]: ax.spines[sp].set_visible(False)
ax.set_xticks([]); ax.tick_params(left=False,labelsize=5)

ax=fig.add_subplot(gs[1,2]); ctry=nf["primary_country"].value_counts().head(8).sort_values()
cl2=["#BBBBBB"]*8; cl2[-1]="#1565C0"
ax.hlines(range(len(ctry)),0,ctry.values,color="#BBBBBB",lw=1.5)
ax.scatter(ctry.values,range(len(ctry)),s=40,c=cl2,zorder=5,edgecolors="white",lw=0.5)
ax.set_yticks(range(len(ctry))); ax.set_yticklabels(ctry.index,fontsize=5)
ax.set_title("(f) Top Countries",fontsize=9,fontweight="bold")
for sp in ["top","right","left"]: ax.spines[sp].set_visible(False)
ax.set_xticks([]); ax.tick_params(left=False)

for a in fig.get_axes(): a.tick_params(labelsize=5); a.spines["top"].set_visible(False); a.spines["right"].set_visible(False)
fig.suptitle(f"Netflix EDA: {len(nf):,} Titles",fontsize=13,fontweight="bold")
plt.savefig("output/W04_M10_netflix.pdf",bbox_inches="tight")
plt.savefig("output/W04_M10_netflix.png",dpi=300,bbox_inches="tight"); plt.close()
print("W04-M10 Lab Complete (Python)")
